src/ucs/datastruct/ptr_map.inl

```diff
@@ -0,0 +1,93 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_PTR_MAP_INL_
+#define UCS_PTR_MAP_INL_
+
+#include "ptr_map.h"
+
+BEGIN_C_DECLS
+
+KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+static inline ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map)
+{
+    map->next_id = 0;
+    kh_init_inplace(ucs_ptr_map_impl, &map->hash);
+    return UCS_OK;
+}
+
+static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
+{
+    kh_destroy_inplace(ucs_ptr_map_impl, &map->hash);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key = (uintptr_t)ptr;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+    ucs_assert_always(!(key & UCS_PTR_MAP_KEY_ID_FLAG));
```

Comment:
why always? we probably don't want this on fast path

Reply:
actually to get unaligned pointer is not so difficult, for example, `char arr[N]`, also vars alignment can be compiler and its options specific, do we really want to take such risk?

Reply:
UCP is allocating the requests and should make sure they are aligned, this check should not be done in release mode, it's extra overhead

Reply:
Ok, will add to docs that all pointers must be aligned on size of pointer. Or maybe some other size?

Reply:
we can define a constant for minimal ptr alignment for ptr_map

Reply:
👍 what value do you prefer?

Reply:
`#define UCS_PTR_MAP_KEY_MINALIGN UCS_BIT(1)` (we can always increase if needed)
and use this value when incrementing next_id